### PR TITLE
Instant Search: Add redux state for API, search, and sort

### PR DESCRIPTION
**Pull request by jsnmoon**
Replaces #18099.

#### Changes proposed in this Pull Request:
* Adds Redux actions, effects, reducers, and selectors for handling API requests, setting user-inputted search queries, and selected sort type.
* The Redux store is not connected to the component tree in this PR; this will occur in a future PR.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Review the code and ensure it looks reasonable.
* Run `yarn test-search` and ensure all tests pass.

#### Proposed changelog entry for your changes:
* None.

### CODE REVIEW

File: modules/search/instant-search/store/selectors.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	return state.sort;
+}
+
+export function getFilters( state ) {
+	return state.filters;
+}
+
+export function hasFilters( state ) {
+	return Object.keys( state.filters ).length > 0;
+}
```
Comment (jsnmoon) [inline modules/search/instant-search/store/selectors.js:31]: These were meant to be added in a follow-up PR and can be safely ignored at this time.

File: modules/search/instant-search/store/actions.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	return {
+		type: 'RECORD_FAILED_SEARCH_REQUEST',
+		error,
+	};
+}
+
+/**
+ * Returns an action object used to initialize query value related reducers.
+ *
+ * @param {string} query - Inputted user query.
```
Comment (bluefuton) [inline modules/search/instant-search/store/actions.js:49]: Is this right? Looks like it accepts an object.

**Review by bluefuton — DISMISSED** [review]
Code looks good, tests pass, and no unexpected side effects on the existing search experience.

One small nit on a code comment but can move straight to Crew review when that is fixed 👍

File: modules/search/instant-search/store/actions.js (hunk 2 above)
Comment (jsnmoon) [reply modules/search/instant-search/store/actions.js:49]: Good eye, fixed!

**Review by jeherve — DISMISSED** [review]
This will need a rebase to fix those unrelated test failures, I'm afraid. It should be good to go after that!

**Comment by jsnmoon** [comment]
Rebased!

**Review by jeherve — APPROVED** [review]
This looks good to me. There are a few ESLint comments here that may be worth addressing to make it easier for others to jump in those files in the future. I would not consider this a blocker though.

**Review by jeherve — APPROVED** [review]
re-approving after a rebase to address unrelated test failures.

**Comment by jsnmoon** [comment]
r219106-wpcom

**Pull request merged**